app/services/db/private-cloud-costs.ts

```diff
@@ -0,0 +1,191 @@
+  }
+
+  changePoints.add(startDate.getTime());
+  changePoints.add(endDate.getTime());
+
+  const sortedPoints = _orderBy(Array.from(changePoints), undefined, 'asc').map((ms) => new Date(ms));
+
+  const costItems: CostItem[] = [];
+
+  for (let i = 0; i < sortedPoints.length - 1; i++) {
+    const intervalStart = sortedPoints[i];
+    const intervalEnd = sortedPoints[i + 1];
+
+    if (intervalEnd <= startDate || intervalStart >= endDate) continue;
+
+    const quota = _last(
+      _filter(
+        allRequests,
+        (
+          req,
+        ): req is (typeof allRequests)[number] & {
+          decisionData: { resourceRequests: Record<string, { cpu: number; storage: number }> };
+        } => !!req.provisionedDate && req.provisionedDate <= intervalStart,
+      ),
+    );
+
+    const price = _last(_filter(unitPrices, (p) => new Date(p.date) <= intervalStart)) ?? {
+      id: 'fallback-zero',
+      cpu: 0,
+      storage: 0,
+      date: intervalStart,
+    };
+
+    if (!quota || !price) continue;
+
+    let totalCpu = 0;
+    let totalStorage = 0;
+
+    if (!quota) continue;
+    const envs = quota.decisionData.resourceRequests;
+    for (const env of namespaceKeys) {
+      const usage = envs[env];
+      if (usage) {
+        totalCpu += usage.cpu || 0;
+        totalStorage += usage.storage || 0;
+      }
+    }
+
+    const durationMinutes = (intervalEnd.getTime() - intervalStart.getTime()) / (1000 * 60);
+    const durationFraction = durationMinutes / (1000 * 60 * 60 * 24);
+    const cpuCost = totalCpu * price.cpu * durationFraction;
+    const storageCost = totalStorage * price.storage * durationFraction;
+    const totalCost = cpuCost + storageCost;
+
+    costItems.push({
+      startDate: intervalStart,
+      endDate: intervalEnd,
+      cpu: totalCpu,
+      storage: totalStorage,
+      cpuCost,
+      storageCost,
+      totalCost,
+      unitPriceId: price.id,
+    });
+  }
+
+  return costItems;
+}
+
+export function estimateMonthTotal(currentTotal: number, year: number, month: number, lastItem?: CostItem) {
+  if (!lastItem) return currentTotal;
+
+  const lastDay = lastItem.endDate.getUTCDate();
+  const daysInMonth = new Date(year, month + 1, 0).getUTCDate();
+
+  const elapsedDays = Math.max(lastDay, 1);
+  const remaining = daysInMonth - elapsedDays;
+
+  const dailyCost = lastItem.totalCost / elapsedDays;
+  return currentTotal + dailyCost * remaining;
+}
+
+export async function getMonthlyCosts(licencePlate: string, year: number, month: number) {
+  const startDate = new Date(Date.UTC(year, month - 1, 1));
+  const endDate = new Date(Date.UTC(year, month, 0, 23, 59, 59, 999));
+
+  const now = new Date();
+  const isCurrentMonth = now.getUTCFullYear() === year && now.getUTCMonth() === month - 1;
+
+  const costItems = await getCostItemsForRange(licencePlate, startDate, endDate);
+
+  let currentTotal: number | undefined = undefined;
```

Comment:
costItems.reduce((sum, i) => sum + i.totalCost, 0);
 ---
 let's make `currentTotal` is always available